src/net/socket_controller.rs

```diff
@@ -0,0 +1,850 @@
+use crate::{
+    config::Config,
+    error::Result,
+    net::{events::SocketEvent, ConnectionController, VirtualConnection},
+    packet::Packet,
+};
+use crossbeam_channel::{self, unbounded, Receiver, Sender};
+use log::error;
+use std::{self, collections::HashMap, fmt::Debug, net::SocketAddr, time::Instant};
+
+/// This trait can be implemented to send data to the socket.
+pub trait SocketSender: Debug {
+    // Send a single packet to the socket.
+    fn send_packet(&mut self, addr: &SocketAddr, payload: &[u8]) -> Result<usize>;
+}
+
+/// This trait can be implemented to receive data from the socket.
+pub trait SocketReceiver: Debug {
+    /// Receives a single packet from the socket.
+    fn receive_packet<'a>(
+        &mut self,
+        buffer: &'a mut [u8],
+    ) -> Result<Option<(&'a [u8], SocketAddr)>>;
+    /// Returns the socket address that this socket was created from.
+    fn local_addr(&self) -> Result<SocketAddr>;
+}
+
+/// A reliable generic socket implementation with configurable reliability and ordering guarantees.
+/// Owns all core components: active connections, connection handler, event sender/receiver, socket sender/receiver.

```

Comment:
Is this a public exported type, so yes we probably don't have to mention what's inside this.

Reply:
Ok, you're right it doesn't convey any useful information.